### PR TITLE
Point to lifetime spans on lifetime errors

### PR DESCRIPTION
**Pull request by estebank**

### CODE REVIEW

File: src/librustc/infer/error_reporting/mod.rs [hunk 1]
```diff
@@ -219,10 +221,30 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
...
         let (prefix, span) = match *region {
-            ty::ReEarlyBound(ref br) => (
-                format!("the lifetime {} as defined on", br.name),
-                self.sess.codemap().def_span(self.hir.span(node)),
-            ),
+            ty::ReEarlyBound(ref br) => {
+                let mut sp = cm.def_span(self.hir.span(node));
+                if let Some(generics) = self.hir.get_generics(scope) {
+                    for param in &generics.params {
+                        if param.name.name().as_str() == br.name.as_str() {
```
Comment (oli-obk) [inline src/librustc/infer/error_reporting/mod.rs]: This loop feels like something that occurs more often (e.g. in lowering code or other diagnostics). Maybe pull it out into a method on `Generics` and do a quick grep around the source to see if any other places could benefit from an `fn(&Generics, Name) -> Option<&Param>`

File: src/librustc/infer/error_reporting/mod.rs [hunk 2]
```diff
@@ -219,10 +221,30 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
             }
         };
         let (prefix, span) = match *region {
-            ty::ReEarlyBound(ref br) => (
-                format!("the lifetime {} as defined on", br.name),
-                self.sess.codemap().def_span(self.hir.span(node)),
-            ),
+            ty::ReEarlyBound(ref br) => {
+                let mut sp = cm.def_span(self.hir.span(node));
+                if let Some(generics) = self.hir.get_generics(scope) {
```
Comment (nikomatsakis) [inline src/librustc/infer/error_reporting/mod.rs]: I agree with @oli-obk that we should not "open code" this loop but rather extract a helper. Bonus points for finding other code to use it =)

File: src/librustc/infer/error_reporting/mod.rs [hunk 3]
```diff
@@ -219,10 +221,30 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
...
+                }
+                (format!("the lifetime {} as defined on", br.name), sp)
+            }
+            ty::ReFree(ty::FreeRegion {
+                bound_region: ty::BoundRegion::BrNamed(_, ref name), ..
+            }) => {
+                let mut sp = cm.def_span(self.hir.span(node));
+                if let Some(generics) = self.hir.get_generics(scope) {
+                    for param in &generics.params {
+                        if param.name.name().as_str() == name.as_str() {
```
Comment (nikomatsakis) [inline src/librustc/infer/error_reporting/mod.rs]: If nothing else, the same loop seems to be down here?

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
Looks good, if we extract a helper would look better. 😛

File: src/librustc/infer/error_reporting/mod.rs (hunk 1 above)
Comment (estebank) [reply src/librustc/infer/error_reporting/mod.rs]: I made a cursory look for other possible uses but I couldn't any. I'm sure I'll find some while working on something else down the line, though.

**Comment by estebank** [comment]
Helper added.

**Review by nikomatsakis — APPROVED** [review]

**Pull request merged**